src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+ * Upon ismaster cancellation, cancelled will be true, and the returned server
+ * description will not contain an error or ismaster reply.
+ * Upon command error ("ok":0 reply), the returned server description have the
+ * ismaster reply and error set.
+ * Returns a new server description that the caller must destroy.
+ */
+mongoc_server_description_t *
+mongoc_server_monitor_check_server (
+   mongoc_server_monitor_t *server_monitor,
+   const mongoc_server_description_t *previous_description,
+   bool *cancelled)
+{
+   bool ret = false;
+   bson_error_t error;
+   bson_t ismaster_reply;
+   int64_t duration_us;
+   int64_t start_us;
+   bool command_or_network_error = false;
+   bool awaited = false;
+   mongoc_server_description_t *description;
+
+   ENTRY;
+
+   *cancelled = false;
+   memset (&error, 0, sizeof (bson_error_t));
+   description = bson_malloc0 (sizeof (mongoc_server_description_t));
+   mongoc_server_description_init (
+      description,
+      server_monitor->description->connection_address,
+      server_monitor->description->id);
+   start_us = _now_us ();
+
+   if (!server_monitor->stream) {
+      MONITOR_LOG (server_monitor, "setting up connection");
+      awaited = false;
+      _server_monitor_heartbeat_started (server_monitor, awaited);
+      ret = _server_monitor_setup_connection (
+         server_monitor, &ismaster_reply, &start_us, &error);
+      GOTO (exit);
+   }
+
+   if (server_monitor->more_to_come) {
+      awaited = true;
+      /* Publish a heartbeat started for each additional response read. */
+      _server_monitor_heartbeat_started (server_monitor, awaited);
+      MONITOR_LOG (server_monitor, "more to come");
+      ret = _server_monitor_awaitable_ismaster_recv (
+         server_monitor, &ismaster_reply, cancelled, &error);
+      GOTO (exit);
+   }
+
+   if (!bson_empty (&previous_description->topology_version)) {
+      awaited = true;
+      _server_monitor_heartbeat_started (server_monitor, awaited);
+      MONITOR_LOG (server_monitor, "awaitable ismaster");
+      ret = _server_monitor_awaitable_ismaster (
+         server_monitor,
+         &previous_description->topology_version,
+         &ismaster_reply,
+         cancelled,
+         &error);
+      GOTO (exit);
+   }
+
+   MONITOR_LOG (server_monitor, "polling ismaster");
+   awaited = false;
+   _server_monitor_heartbeat_started (server_monitor, awaited);
+   ret = _server_monitor_polling_ismaster (
+      server_monitor, &ismaster_reply, &error);
+
+exit:
+   duration_us = _now_us () - start_us;
+   MONITOR_LOG (
+      server_monitor, "server check duration (us): %" PRId64, duration_us);
+
+   if (ret) {
+      int64_t rtt_ms;
+      /* rtt remains -1 if awaited. */
+      if (awaited) {
+         rtt_ms = -1;
+      } else {
+         rtt_ms = duration_us / 1000;
+      }
+      mongoc_server_description_handle_ismaster (
+         description, &ismaster_reply, rtt_ms, NULL);
+      /* Check if it is a command error ("ok":0). */
```

Comment:
Should this actually check `ok`? As-is, you're just looking at `error.code`. Feel free to clarify is that's simply the way that `mongoc_server_description_handle_ismaster` relays that `ok` was zero -- in which case it probably makes sense to change this comment to something like `error.code will be non-zero on command error ("ok":0)`